lib/backend/k8s/syncer.go

```diff
@@ -0,0 +1,536 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/compat"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	k8sapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/watch"
+	"reflect"
+)
+
+func newSyncer(kc KubeClient, callbacks api.SyncerCallbacks) *kubeSyncer {
+	return &kubeSyncer{
+		kc:        kc,
+		callbacks: callbacks,
+	}
+}
+
+type kubeSyncer struct {
+	kc        KubeClient
+	callbacks api.SyncerCallbacks
+	OneShot   bool
+}
+
+// Holds resource version information.
+type resourceVersions struct {
+	podVersion           string
+	namespaceVersion     string
+	networkPolicyVersion string
+}
+
+func getUpdateType(e watch.Event) model.UpdateType {
+	// Determine the update type.
+	updateType := model.UpdateTypeKVUnknown
+	switch e.Type {
+	case watch.Added:
+		updateType = model.UpdateTypeKVNew
+	case watch.Deleted:
+		updateType = model.UpdateTypeKVDeleted
+	case watch.Modified:
+		updateType = model.UpdateTypeKVUpdated
+	}
+	return updateType
+}
+
+func (syn *kubeSyncer) Start() {
+	// Channel for receiving updates from a snapshot.
+	snapshotUpdates := make(chan *[]model.Update)
+
+	// Channel for receiving updates from the watcher.
+	watchUpdates := make(chan *model.Update)
+
+	// Channel used by the watcher to trigger a re-sync.
+	triggerResync := make(chan *resourceVersions, 5)
```

Comment:
I recommend switching these from `*resourceVersions` to `resourceVersions` to rule out bugs caused by accidentally re-using the struct after sending it on the channel.